<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Evan Paull and Sunny Jones" />


<title>MOMA - Multi Omic Master Regulator Analysis</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#header {
text-align: center;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; }  code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">MOMA - Multi Omic Master Regulator Analysis</h1>
<h4 class="author">Evan Paull and Sunny Jones</h4>



<div id="introduction-and-background" class="section level1">
<h1>Introduction and Background</h1>
<p>MOMA is a tool for inferring connections between Master Regulator proteins and genomic driver events in cancer. Master regulators are regulatory proteins (primarily transcription factors and co-transcription factors) that control cell state. In the case of cancer and other disease states transcription factors have been shown to be key drivers of maintaining the disease state and can be targets for interventions. Often these master regulators are not mutated themselves but are downstream of mutations and other genomic alterations that ultimately dysregulate the normal activity of that regulator. MOMA uses multiple inputs of information to infer these connections and to improve the predictive value of the master regulator analysis.</p>
<p><em>For more information on Master Regulators and our tool for calculating their values see the paper published on VIPER in Nature Genetics in 2016</em> <a href="10.1038/ng.3593">paper</a> // <a href="10.18129/B9.bioc.viper">package</a>.</p>
<hr />
</div>
<div id="running-moma" class="section level1">
<h1>Running MOMA</h1>
<div id="getting-started" class="section level2">
<h2>Getting Started</h2>
<p>First load the library into the R session. Make sure all the other dependent packages have already been installed to ensure full functionality, including graphics and plotting functions.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw">library</span>(moma)</a></code></pre></div>
<p>Explore the test data, and confirm that we have 100 test samples and 2506 VIPER inferred proteins in the test viper matrix. This was generated by using VIPER on the GBM gene expression signature. The matrix has the samples across the columns and the regulators in the rows.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw">names</span>(gbm.example)</a>
<a class="sourceLine" id="cb2-2" data-line-number="2"><span class="co">#&gt;  [1] &quot;vipermat&quot;         &quot;cindy&quot;            &quot;preppi&quot;          </span></a>
<a class="sourceLine" id="cb2-3" data-line-number="3"><span class="co">#&gt;  [4] &quot;rawsnp&quot;           &quot;fusions&quot;          &quot;gene.loc.mapping&quot;</span></a>
<a class="sourceLine" id="cb2-4" data-line-number="4"><span class="co">#&gt;  [7] &quot;clinical&quot;         &quot;fCNV&quot;             &quot;mutSig&quot;          </span></a>
<a class="sourceLine" id="cb2-5" data-line-number="5"><span class="co">#&gt; [10] &quot;rawcnv&quot;</span></a>
<a class="sourceLine" id="cb2-6" data-line-number="6"><span class="kw">dim</span>(gbm.example<span class="op">$</span>vipermat)</a>
<a class="sourceLine" id="cb2-7" data-line-number="7"><span class="co">#&gt; [1] 2506  100</span></a>
<a class="sourceLine" id="cb2-8" data-line-number="8">gbm.example<span class="op">$</span>vipermat[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]</a>
<a class="sourceLine" id="cb2-9" data-line-number="9"><span class="co">#&gt;     TCGA-14-1825-01 TCGA-76-4931-01 TCGA-06-5418-01</span></a>
<a class="sourceLine" id="cb2-10" data-line-number="10"><span class="co">#&gt; 165      -3.0369845       0.9518478       1.1485397</span></a>
<a class="sourceLine" id="cb2-11" data-line-number="11"><span class="co">#&gt; 196      -2.4667300      -3.2473674      -0.5219556</span></a>
<a class="sourceLine" id="cb2-12" data-line-number="12"><span class="co">#&gt; 257       0.7872116       0.7562328       0.6428189</span></a></code></pre></div>
<p>Next, we select the biological pathways we’d like to use to aid in our inference task. In this case, we’re using the CINDy modulator inference algorithm [CITE] as well as protein-protein interactions predicted by the PrePPI structure-based algorithm [CITE]. The output of the CINDy algorithm is a likelihood (p value) for the association of an upstream modulator with a particular regulators activity. The relevant output for PrePPI is a likelihood (p value) that a modulator structurally binds to a regulator.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1">pathways &lt;-<span class="st"> </span><span class="kw">list</span>()</a>
<a class="sourceLine" id="cb3-2" data-line-number="2">pathways[[<span class="st">'cindy'</span>]] =<span class="st"> </span>gbm.example<span class="op">$</span>cindy</a>
<a class="sourceLine" id="cb3-3" data-line-number="3">pathways[[<span class="st">'preppi'</span>]] =<span class="st"> </span>gbm.example<span class="op">$</span>preppi</a></code></pre></div>
</div>
<div id="generating-the-moma-object" class="section level2">
<h2>Generating the <strong>MOMA</strong> object</h2>
<p>Generate the MOMA object from the data that we have. The data required to do the full analysis are:</p>
<ul>
<li><strong>VIPER matrix</strong>: continuous values of inferred protein activities</li>
<li><strong>SNP matrix</strong>: binary 0/1 values where 1 indicates the prescence of a mutation</li>
<li><strong>CNV matrix</strong>: continuous values as calculated by GISTIC [CITE]</li>
<li><strong>Fusion matrix</strong>: binary 0/1 values</li>
<li><strong>Pathway list</strong>: see above</li>
<li><strong>Gene Blacklist</strong>: genes to not include in the analysis *</li>
<li><strong>Gene Locations</strong>: mapping of genes to chromosome locations</li>
<li><strong>Output Folder</strong> : (optional) will save intermediate DIGGIT results here</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1">momaObj &lt;-<span class="st"> </span><span class="kw">moma.constructor</span>(gbm.example<span class="op">$</span>vipermat, gbm.example<span class="op">$</span>rawsnp,</a>
<a class="sourceLine" id="cb4-2" data-line-number="2">        gbm.example<span class="op">$</span>rawcnv, gbm.example<span class="op">$</span>fusions, pathways,</a>
<a class="sourceLine" id="cb4-3" data-line-number="3">        <span class="dt">gene.blacklist=</span>gbm.example<span class="op">$</span>mutSig,</a>
<a class="sourceLine" id="cb4-4" data-line-number="4">        <span class="dt">gene.loc.mapping=</span>gbm.example<span class="op">$</span>gene.loc.mapping)</a>
<a class="sourceLine" id="cb4-5" data-line-number="5"><span class="co">#&gt; [1] &quot;Number of samples in VIPER + Mutation data: 100&quot;</span></a>
<a class="sourceLine" id="cb4-6" data-line-number="6"><span class="co">#&gt; [1] &quot;Number of samples in VIPER + CNV data: 100&quot;</span></a>
<a class="sourceLine" id="cb4-7" data-line-number="7"><span class="co">#&gt; [1] &quot;Number of samples in VIPER + Fusion data: 68&quot;</span></a>
<a class="sourceLine" id="cb4-8" data-line-number="8"><span class="co">#&gt; [1] &quot;Common samples with all data analysis:  100&quot;</span></a>
<a class="sourceLine" id="cb4-9" data-line-number="9"><span class="co">#&gt; [1] &quot;Checking labels on pathway cindy&quot;</span></a>
<a class="sourceLine" id="cb4-10" data-line-number="10"><span class="co">#&gt; [1] &quot;Found labels for  364  TFs in VIPER matrix&quot;</span></a>
<a class="sourceLine" id="cb4-11" data-line-number="11"><span class="co">#&gt; [1] &quot;Checking labels on pathway preppi&quot;</span></a>
<a class="sourceLine" id="cb4-12" data-line-number="12"><span class="co">#&gt; [1] &quot;Found labels for  2506  TFs in VIPER matrix&quot;</span></a></code></pre></div>
</div>
<div id="moma-analysis-on-gbm-data" class="section level2">
<h2>MOMA Analysis on GBM Data</h2>
<p>The first step, <code>runDIGGIT()</code> will run the DIGGIT inference algorithm [CITE] to findstatistical interactions between VIPER-inferred proteins and genomic events.</p>
<p>The <code>makeInteractions()</code> function will infer robust computational predictions using all the provided data, including the Conditional Inference of Network Dynamics (CINDy) algorithm.<br />
The option <code>cindy.only=FALSE</code> will allow interactions without evidence from CINDy, but with strong evidence from other sources, to be used.</p>
<p>The <code>Rank()</code> function will create a final ranking of candidate Master Regulators for this cohort of patient samples.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1">momaObj<span class="op">$</span><span class="kw">runDIGGIT</span>(<span class="dt">fCNV=</span>gbm.example<span class="op">$</span>fCNV)</a>
<a class="sourceLine" id="cb5-2" data-line-number="2"></a>
<a class="sourceLine" id="cb5-3" data-line-number="3">momaObj<span class="op">$</span><span class="kw">makeInteractions</span>(<span class="dt">cindy.only=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb5-4" data-line-number="4"></a>
<a class="sourceLine" id="cb5-5" data-line-number="5">momaObj<span class="op">$</span><span class="kw">Rank</span>(<span class="dt">use.cindy=</span><span class="ot">TRUE</span>)</a></code></pre></div>
<p>Clustering of the samples, using the protein ranks computed in the last step, can then be performed using <code>Cluster()</code>. Multiple cluster solutions will be calculated, ranging from 2 to 10 clusters by default. The silhouette or reliability score of each can be assessed to determine an optimal ‘k’ number of clusters; here, we are selecting the 3 cluster solution. <em>(Because there are no 1 cluster solutions, the index of the desired solution in the generated clustering object is k-1).</em></p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1">clustering.solutions &lt;-<span class="st"> </span>momaObj<span class="op">$</span><span class="kw">Cluster</span>()</a>
<a class="sourceLine" id="cb6-2" data-line-number="2"><span class="co">#&gt; [1] &quot;using pearson correlation with weighted vipermat&quot;</span></a>
<a class="sourceLine" id="cb6-3" data-line-number="3"><span class="co">#&gt; [1] &quot;testing clustering options, k = 2..15&quot;</span></a>
<a class="sourceLine" id="cb6-4" data-line-number="4"><span class="co">#&gt; [1] &quot;2 0.618073073073074&quot;</span></a>
<a class="sourceLine" id="cb6-5" data-line-number="5"><span class="co">#&gt; [1] &quot;3 0.818053053053053&quot;</span></a>
<a class="sourceLine" id="cb6-6" data-line-number="6"><span class="co">#&gt; [1] &quot;4 0.75460960960961&quot;</span></a>
<a class="sourceLine" id="cb6-7" data-line-number="7"><span class="co">#&gt; [1] &quot;5 0.79466966966967&quot;</span></a>
<a class="sourceLine" id="cb6-8" data-line-number="8"><span class="co">#&gt; [1] &quot;6 0.765690690690691&quot;</span></a>
<a class="sourceLine" id="cb6-9" data-line-number="9"><span class="co">#&gt; [1] &quot;7 0.781246246246247&quot;</span></a>
<a class="sourceLine" id="cb6-10" data-line-number="10"><span class="co">#&gt; [1] &quot;8 0.759434434434434&quot;</span></a>
<a class="sourceLine" id="cb6-11" data-line-number="11"><span class="co">#&gt; [1] &quot;9 0.739864864864865&quot;</span></a>
<a class="sourceLine" id="cb6-12" data-line-number="12"><span class="co">#&gt; [1] &quot;10 0.652727727727728&quot;</span></a>
<a class="sourceLine" id="cb6-13" data-line-number="13"><span class="co">#&gt; [1] &quot;11 0.702447447447448&quot;</span></a>
<a class="sourceLine" id="cb6-14" data-line-number="14"><span class="co">#&gt; [1] &quot;12 0.697382382382383&quot;</span></a>
<a class="sourceLine" id="cb6-15" data-line-number="15"><span class="co">#&gt; [1] &quot;13 0.626146146146146&quot;</span></a>
<a class="sourceLine" id="cb6-16" data-line-number="16"><span class="co">#&gt; [1] &quot;14 0.565650650650651&quot;</span></a>
<a class="sourceLine" id="cb6-17" data-line-number="17"><span class="co">#&gt; [1] &quot;15 0.569354354354355&quot;</span></a>
<a class="sourceLine" id="cb6-18" data-line-number="18"></a>
<a class="sourceLine" id="cb6-19" data-line-number="19"><span class="co"># pick the 3 cluster solution and save it to the moma Object</span></a>
<a class="sourceLine" id="cb6-20" data-line-number="20">momaObj<span class="op">$</span>sample.clustering &lt;-<span class="st"> </span>clustering.solutions[[<span class="dv">2</span>]]<span class="op">$</span>clustering</a></code></pre></div>
<p>Genomic saturation analysis is then performed on each cluster with the <code>saturationPlots()</code> function, allowing us to find the key proteins that are downstream of the majority of genomic events in the samples within a particular cluster. These regulators make up that cluster’s Checkpoint.</p>
<p>The results of this analysis can be accessed directly in the following result fields:</p>
<ul>
<li><code>momaObj$checkpoints</code></li>
<li><code>momaObj$genomic.saturation</code></li>
<li><code>momaObj$coverage.summaryStats</code></li>
</ul>
<p>These will be used for plotting the genomic saturation curves as well.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1">momaObj<span class="op">$</span><span class="kw">saturationPlots</span>()</a></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1">cluster1.checkpoint &lt;-<span class="st"> </span>momaObj<span class="op">$</span>checkpoints[[<span class="dv">1</span>]]</a>
<a class="sourceLine" id="cb8-2" data-line-number="2"><span class="kw">print</span> (cluster1.checkpoint[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>])</a>
<a class="sourceLine" id="cb8-3" data-line-number="3"><span class="co">#&gt;  [1] &quot;148103&quot; &quot;9817&quot;   &quot;84775&quot;  &quot;10224&quot;  &quot;23660&quot;  &quot;6662&quot;   &quot;57106&quot; </span></a>
<a class="sourceLine" id="cb8-4" data-line-number="4"><span class="co">#&gt;  [8] &quot;147741&quot; &quot;26147&quot;  &quot;57615&quot;</span></a></code></pre></div>
<hr />
</div>
</div>
<div id="plotting-the-results" class="section level1">
<h1>Plotting the Results</h1>
<div id="visualizing-the-viper-matrix" class="section level2">
<h2>Visualizing the VIPER matrix</h2>
<p>The primary results of the analysis are the master regululators of each particular cluster’s checkpoint, as displayed above. You can plot the original VIPER matrix subset down to only these regulators using whatever heatmap function you like. See example below generated using ComplexHeatmap [CITE]<br />
[IMAGE].</p>
</div>
<div id="visualizing-the-genomic-saturation-and-events-plots" class="section level2">
<h2>Visualizing the Genomic Saturation and Events Plots</h2>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
