---
title: "MOMA - Multi Omic Master Regulator Analysis"
author: "Evan Paull and Sunny Jones"
output: rmarkdown::html_vignette
#bibliography: bibliography.bib
vignette: >
  %\VignetteIndexEntry{MOMA - Multi Omic Master Regulator Analysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Introduction and Background

MOMA is a tool for inferring connections between Master Regulator proteins and 
genomic driver events in cancer. Master regulators are regulatory proteins 
(primarily transcription factors and co-transcription factors) that control cell
state. In the case of cancer and other disease states transcription factors have 
been shown to be key drivers of maintaining the disease state and can be targets 
for interventions. Often these master regulators are not mutated themselves but 
are downstream of mutations and other genomic alterations that ultimately 
dysregulate the normal activity of that regulator. MOMA uses multiple inputs of 
information to infer these connections and to improve the predictive value of 
the master regulator analysis.  



*For more information on Master Regulators and our tool for calculating their values see the paper published on VIPER in Nature Genetics in 2016* 
[paper](10.1038/ng.3593) // [package](10.18129/B9.bioc.viper).

*** 

# Running MOMA

## Getting Started

First load the library into the R session. Make sure all the other dependent 
packages have already been installed to ensure full functionality, including 
graphics and plotting functions.  


```{r setup}
knitr::opts_chunk$set(warning = FALSE, message = FALSE)
library(moma)
```


Explore the test data, and confirm that we have 100 test samples and 2506 VIPER 
inferred proteins in the test viper matrix. This was generated by using VIPER on 
the GBM gene expression signature. The matrix has the samples across the columns 
and the regulators in the rows.  


```{r explore data}
names(gbm.example)
dim(gbm.example$vipermat)
gbm.example$vipermat[1:3, 1:3]
```

Next, we select the biological pathways we'd like to use to aid in our inference 
task. In this case, we're using the CINDy modulator inference algorithm [CITE] 
as well as protein-protein interactions predicted by the PrePPI structure-based 
algorithm [CITE]. The output of the CINDy algorithm is a likelihood (p value) for
the association of an upstream modulator with a particular regulators activity. 
The relevant output for PrePPI is a likelihood (p value) that a modulator 
structurally binds to a regulator.  

```{r pathways}
pathways <- list()
pathways[['cindy']] = gbm.example$cindy
pathways[['preppi']] = gbm.example$preppi
```

## Generating the **MOMA** object

Start by loading in all the data we have to a MOMA object. The following are data required 
to do the full analysis:

- **VIPER matrix**: continuous values of inferred protein activities
- **SNP matrix**: binary 0/1 values where 1 indicates the prescence of a mutation
- **CNV matrix**: continuous values as calculated by GISTIC [CITE]
- **Fusion matrix**: binary 0/1 values 
- **Pathway list**: see above
- **Gene Blacklist**: genes to not include in the analysis *
- **Gene Locations**: mapping of genes to chromosome locations
- **Output Folder** : (optional) will save intermediate DIGGIT results here

```{r moma object}
momaObj <- moma.constructor(gbm.example$vipermat, gbm.example$rawsnp,
        gbm.example$rawcnv, gbm.example$fusions, pathways,
        gene.blacklist=gbm.example$mutSig,
        gene.loc.mapping=gbm.example$gene.loc.mapping)
```

## MOMA Analysis on GBM Data

The first step, `runDIGGIT()`  will run the DIGGIT inference algorithm [CITE] to
find statistical interactions between VIPER-inferred proteins and genomic events.  

The `makeInteractions()` function will infer robust computational predictions
using all the provided data, including the Conditional Inference of Network Dynamics 
(CINDy) algorithm.  
The option `cindy.only=FALSE` will allow interactions without evidence from CINDy,
but with strong evidence from other sources, to be used.  

The `Rank()` function will create a final ranking of candidate Master Regulators
for this cohort of patient samples.

```{r interactions, results = "hide", message = FALSE}
momaObj$runDIGGIT(fCNV=gbm.example$fCNV)

momaObj$makeInteractions(cindy.only=FALSE)

momaObj$Rank(use.cindy=TRUE)
```

Clustering of the samples, using the protein ranks computed in the last step, 
can then be performed using `Cluster()`. Multiple cluster solutions will be 
calculated, ranging from 2 to 10 clusters by default. The silhouette or 
reliability score of each can be assessed to determine an optimal 'k' number of 
clusters; here, we are selecting the 3 cluster solution. 
*(Because there are no 1 cluster solutions, the index of the desired solution 
in the generated clustering object is k-1).* 

```{r clustering}
clustering.solutions <- momaObj$Cluster()

# pick the 3 cluster solution and save it to the moma Object
momaObj$sample.clustering <- clustering.solutions[[2]]$clustering
```

  
Genomic saturation analysis is then performed on each cluster with the 
`saturationCalculation()` function, allowing us to find the key proteins that are 
downstream of the majority of genomic events in the samples within a particular 
cluster. These regulators make up that cluster's Checkpoint.  

The results of this analysis can be accessed directly in the following result 
fields:  
  
  - `momaObj$checkpoints` 
  - `momaObj$genomic.saturation`
  - `momaObj$coverage.summaryStats` 
  
These will be used for plotting the genomic saturation curves as well.
  
```{r saturation, message=FALSE, warning=FALSE, results="hide"}
momaObj$saturationCalculation()
```

```{r checkpoints}
cluster1.checkpoint <- momaObj$checkpoints[[1]]
print (cluster1.checkpoint[1:10])
```

***

# Plotting the Results

## Visualizing the VIPER matrix  

The primary results of the analysis are the master regululators of each particular
cluster's checkpoint, as displayed above. You can plot the original VIPER matrix 
subset down to only these regulators using whatever heatmap function you like. 
See example below generated using ComplexHeatmap [CITE]  
[IMAGE].

## Visualizing the Genomic Saturation and Events Plots

The other important results of the analysis are the statisically significant
genomic events found to be upstream of each of the checkpoint master regulators. 
The `makeSaturationPlots()` function takes the subtype specific genomic event interactions
calculated in the previous step and makes customizable plots for viewing the results. These
are saved to the $identity.plots() part of the momaObj.

```{r make plots, warning=FALSE}

momaObj$makeSaturationPlots()

```


The two plot types are: 

- **Genomic Event Bar Plots** : these plot the frequency of the different types of 
genomic events found to be upstream of the regulators in the checkpoint.  
- **Saturation Curves** : these plot the number of regulators needed to capture
a certain percentage of genomic events. The inflection points plotted on
this graph match up with the cut off for which regulators make up a particular
cluster's checkpoint.   

```{r visualize plots, fig.height=5, fig.width=7}

momaObj$identity.plots$bar.plots[[1]]

momaObj$identity.plots$curve.plots[[1]]

```

The data for these plots are stored as ggplot objects so layers can be added afterwards
for further customization.

```{r plot customization, fig.height=5, fig.width=7}
library(ggplot2)
momaObj$identity.plots$curve.plots[[1]] +
  ggtitle("GBM Subtype 1") +
  theme_light()

```


## Incorporating and Plotting Survival Data





